compiler/rustc_lint/src/builtin.rs

```diff
@@ -818,25 +818,22 @@ impl EarlyLintPass for DeprecatedAttr {
     fn check_attribute(&mut self, cx: &EarlyContext<'_>, attr: &ast::Attribute) {
         for BuiltinAttribute { name, gate, .. } in &self.depr_attrs {
             if attr.ident().map(|ident| ident.name) == Some(*name) {
-                if let &AttributeGate::Gated(
-                    Stability::Deprecated(link, suggestion),
-                    name,
-                    reason,
-                    _,
-                ) = gate
+                if let &AttributeGate::Gated {
+                    stability: Stability::Deprecated(link, suggestion),
+                    message,
+                    ..
+                } = gate
                 {
                     let suggestion = match suggestion {
-                        Some(msg) => {
-                            BuiltinDeprecatedAttrLinkSuggestion::Msg { suggestion: attr.span, msg }
-                        }
-                        None => {
-                            BuiltinDeprecatedAttrLinkSuggestion::Default { suggestion: attr.span }
+                        Some(suggestion) => {
+                            BuiltinDeprecatedAttrLinkSuggestion::Msg { span: attr.span, suggestion }
                         }
+                        None => BuiltinDeprecatedAttrLinkSuggestion::Default { span: attr.span },
                     };
                     cx.emit_span_lint(
                         DEPRECATED,
                         attr.span,
-                        BuiltinDeprecatedAttrLink { name, reason, link, suggestion },
+                        BuiltinDeprecatedAttrLink { name: *name, message, link, suggestion },
```

Comment:
It looks like this code had some shadowing issues causing `name` to refer to the feature gate name, not the attribute. So I've changed it back, but note that I think this function is dead code.

Reply:
Hmmmmm. Why? If this is dead code then something is wrong. Could we add a test for deprecated attributes?

Reply:
> Hmmmmm. Why? If this is dead code then something is wrong. Could we add a test for deprecated attributes?

There are no deprecated attributes. Actually, `#[macro_escape]` is deprecated but it's not marked at such because it has its own diagnostic in [`rustc_resolve`](https://github.com/rust-lang/rust/blob/d9a739363878a06633eb912e44f02a3c5d6ba08c/compiler/rustc_resolve/src/build_reduced_graph.rs#L1149) instead. This error has a branch on whether it's used as an inner or outer attribute, which the rustc_feature machinery does not support.

Rather than figure out some way to test this, I'd prefer to delete `rustc_feature::Stability` and this `DeprecatedAttr` lint pass instead, and have deprecations be done in check_attr.rs or in the new attribute parsing infra instead.